synthesizer/program/src/logic/instruction/operation/cast.rs

```diff
@@ -381,57 +395,16 @@ impl<N: Network, const VARIANT: u8> CastOperation<N, VARIANT> {
-                }
-
-                // Retrieve the struct and ensure it is defined in the program.
-                let struct_ = stack.program().get_struct(struct_)?;
-
-                // Ensure that the number of operands is equal to the number of struct members.
-                if inputs.len() != struct_.members().len() {
-                    bail!(
-                        "Casting to the struct {} requires {} operands, but {} were provided",
-                        struct_.name(),
-                        struct_.members().len(),
-                        inputs.len()
-                    )
-                }
-
-                // Initialize the struct members.
-                let mut members = IndexMap::new();
-                for (member, (member_name, member_type)) in inputs.iter().zip_eq(struct_.members()) {
-                    // Retrieve the plaintext value from the entry.
-                    let plaintext = match member {
-                        circuit::Value::Plaintext(plaintext) => {
-                            // Ensure the member matches the register type.
-                            stack.matches_plaintext(&plaintext.eject_value(), member_type)?;
-                            // Output the plaintext.
-                            plaintext.clone()
-                        }
-                        // Ensure the struct member is not a record.
-                        circuit::Value::Record(..) => {
-                            bail!("Casting a record into a struct member is illegal")
-                        }
-                        // Ensure the struct member is not a future.
-                        circuit::Value::Future(..) => {
-                            bail!("Casting a future into a struct member is illegal")
-                        }
-                    };
-                    // Append the member to the struct members.
-                    members.insert(circuit::Identifier::constant(*member_name), plaintext);
-                }
-
-                // Construct the struct.
-                let struct_ = circuit::Plaintext::Struct(members, Default::default());
+            CastType::Plaintext(PlaintextType::ExternalStruct(locator)) => {
+                let external_stack = stack.get_external_stack(locator.program_id())?;
+                let plaintext = self.execute_cast_to_struct(&*external_stack, *locator.resource(), inputs)?;
+                // Store the struct.
+                registers.store_circuit(stack, &self.destination, plaintext.into())
+            }
+            CastType::Plaintext(PlaintextType::Struct(struct_name)) => {
+                let plaintext = self.execute_cast_to_struct(stack, *struct_name, inputs)?;
                 // Store the struct.
-                registers.store_circuit(stack, &self.destination, circuit::Value::Plaintext(struct_))
+                registers.store_circuit(stack, &self.destination, plaintext.into())
```

Comment:
Please double check this logic. It is now possible to cast data to an external struct in our circuit. But because those external structs are static and unaffected by program upgradability, no issues can occur.

Double clicking on this because program upgradability audit found issues with the way we handled **values** from external programs in the circuit.